File: src/core/cmd/events/gamepad.rs
use serde::{Deserialize, Serialize};

use super::common::ElementState;

/// Gamepad button types following standard gamepad mapping
#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum GamepadButton {
    // Face buttons
    South = 0, // A / Cross
    East = 1,  // B / Circle
    West = 2,  // X / Square
    North = 3, // Y / Triangle

    // Shoulder buttons
    LeftBumper = 4,
    RightBumper = 5,
    LeftTrigger = 6,
    RightTrigger = 7,

    // Center buttons
    Select = 8,
    Start = 9,
    Mode = 10, // Guide / Home

    // Stick buttons
    LeftStick = 11,
    RightStick = 12,

    // D-pad
    DpadUp = 13,
    DpadDown = 14,
    DpadLeft = 15,
    DpadRight = 16,

    // Other
    Other(u8),
}

/// Gamepad axis types
#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum GamepadAxis {
    LeftStickX = 0,
    LeftStickY = 1,
    RightStickX = 2,
    RightStickY = 3,
    LeftTrigger = 4,
    RightTrigger = 5,
    Other(u8),
}

/// Gamepad events
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "event", content = "data", rename_all = "kebab-case")]
pub enum GamepadEvent {
    /// Gamepad was connected
    OnConnect { gamepad_id: u32, name: String },

    /// Gamepad was disconnected
    OnDisconnect { gamepad_id: u32 },

    /// Gamepad button pressed/released
    OnButton {
        gamepad_id: u32,
        button: GamepadButton,
        state: ElementState,
        value: f32, // 0.0-1.0 for analog triggers
    },

    /// Gamepad axis moved
    OnAxis {
        gamepad_id: u32,
        axis: GamepadAxis,
        value: f32, // -1.0 to 1.0 for sticks, 0.0 to 1.0 for triggers
    },
}

File: binding/src/events/gamepad.ts
import type { ElementState } from './common';

// MARK: Gamepad Events

/** Gamepad button types following standard gamepad mapping */
export type GamepadButton =
  // Face buttons
  | 'south' // A / Cross
  | 'east' // B / Circle
  | 'west' // X / Square
  | 'north' // Y / Triangle
  // Shoulder buttons
  | 'left-bumper'
  | 'right-bumper'
  | 'left-trigger'
  | 'right-trigger'
  // Center buttons
  | 'select'
  | 'start'
  | 'mode' // Guide / Home
  // Stick buttons
  | 'left-stick'
  | 'right-stick'
  // D-pad
  | 'dpad-up'
  | 'dpad-down'
  | 'dpad-left'
  | 'dpad-right'
  // Other
  | { other: number };

/** Gamepad axis types */
export type GamepadAxis =
  | 'left-stick-x'
  | 'left-stick-y'
  | 'right-stick-x'
  | 'right-stick-y'
  | 'left-trigger'
  | 'right-trigger'
  | { other: number };

export interface GamepadConnectedEvent {
  event: 'on-connect';
  data: {
    gamepadId: number;
    name: string;
  };
}

export interface GamepadDisconnectedEvent {
  event: 'on-disconnect';
  data: { gamepadId: number };
}

export interface GamepadButtonEvent {
  event: 'on-button';
  data: {
    gamepadId: number;
    button: GamepadButton;
    state: ElementState;
    value: number; // 0.0-1.0 for analog triggers
  };
}

export interface GamepadAxisEvent {
  event: 'on-axis';
  data: {
    gamepadId: number;
    axis: GamepadAxis;
    value: number; // -1.0 to 1.0 for sticks, 0.0 to 1.0 for triggers
  };
}

export type GamepadEvent =
  | GamepadConnectedEvent
  | GamepadDisconnectedEvent
  | GamepadButtonEvent
  | GamepadAxisEvent;
